### ISSUE TITLE
JS type-checking suggested refactoring produces maximum possible redundant relative module path

### ISSUE THREAD

**Issue by axefrog** [Microsoft/TypeScript#29194]
I'm trying to use JavaScript type-checking (which obviously is powered by TypeScript) in my project, and am noticing that it seems unable to correctly produce minimal relative paths when executing the "Infer parameter types from usage" refactoring.

**TypeScript Version:**  3.2.2 and 3.3.0-dev.20181228

**Search Terms:**
javascript checkjs module path infer relative

**Code**

jsconfig.json (in root of folder):

```json
{
  "compilerOptions": {
    "checkJs": true,
    "target": "esnext",
    "module": "commonjs",
  },
  "include": [
    "src/**/*.js"
  ]
}
```

```js
// src/foo.js
export class Foo {}

// src/bar.js
export function run (foo) { }

// src/main.js
import { run } from './bar';
import { Foo } from './foo';

function main () {
  run(new Foo());
}
```

**Expected behavior:**

In bar.js, the type information suggested refactorings includes "Infer parameter types from usage". Upon performing this refactoring, the `foo` parameter should become associated with a minimal path to the target module, like so:

```js
/**
 * @param {import("./foo").Foo} foo
 */
export function run (foo) { }
```

**Actual behavior:**

Instead, it generates this:

```js
/**
 * @param {import("../../../../../../Dropbox/Work/Dev/salix/current/src/foo").Foo} foo
 */
export function run (foo) { }
```

I have tested this in both the stable and insider builds of VS Code. I have tried various different combinations of settings in jsconfig.json, including `include`, `exclude`, `compilerOptions.baseUrl`, `compilerOptions.module`, and so on. The issue persists with every permutation of options I've tried.

**Related Issues:** 
No relevant issues found in either of the TypeScript or VSCode repositories.

**Comment by fatcerberus** [comment]
> `../../../../../../Dropbox/Work/Dev/salix/current/src/foo`

😮

That's crazy, but also kind of impressive.  It's like it just gives up trying to create a relative path and goes all the way back to the root, but then changes its mind halfway through and goes ahead making a relative path out of *that*.

**Comment by axefrog** [comment]
@fatcerberus And it's kind of weird considering the fact that it knows where the imported symbol lives, and where it's being imported. Why wouldn't it generate a minimal relative path between the two by default?

**Comment by fatcerberus** [comment]
It clearly *does* know how to get from one location to the other based on the fact the final path is still relative, but it seems to be taking the longest route possible to get there.  It's very bizarre, I agree.

**Comment by fatcerberus** [comment]
I wonder: Are there any directory symlinks in the way that might make it have to take a longer route?  Just a hunch, and I don't actually even know if it has logic to account for that, but it's something that I thought of as a potential cause.

**Comment by axefrog** [comment]
@fatcerberus No, actually I just realised you misinterpreted the path example I provided - my Dropbox folder is at the root of the drive where the code is stored. The algorithm doesn't therefore change its mind partway through the process; it just reaches a point where it can't go any deeper, due to having arrived at the drive root. In effect, it's generating an absolute relative path, which obviously is kind of pointless.

**Comment by fatcerberus** [comment]
That’s exactly what I meant, it’s like when a relative path algorithm decides it can’t generate a relative path so switches to absolute mode, so it did that, and then changed its mind and made a relative path anyway—by way of the root directory.

**Comment by weswigham** [comment]
What platform are you on and what's the case-sensitivity of the host filesystem (and what drive letter if windows)?

If I has to guess, I'd assume that somewhere we're doing a case-sensitive path component comparison when we should be doing a case-insensitive one.

**Comment by fatcerberus** [comment]
I know that in some cases (haven’t narrowed down which ones), on Windows, VSCode can end up a mixed-case path is all lowercase (including the drive letter), so that’s my best guess as well.

**Comment by axefrog** [comment]
I'm on Windows 10, drive D, and my literal path strings match the casing stored on the drive.

**Comment by sheetalkamat** [comment]
Getting us tsserver log using [typescript nightly drop](https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/Nightly%20Builds.md#visual-studio-code) might help with the investigation.
You can upload logs by following these steps:
1. With "typescript.tsserver.log": "verbose" enabled
2. Restart vscode or vscode insiders
3. Reproduce the problem
4. You can share the logs using
    - either by sharing logs in the folder that gets opened by running `open tsserver log` from command palette 
    or
    - Run ```code --upload-logs``` on the command line (or ```code-insiders --upload-logs``` if using vscode insiders) and follow instructions and Then share the log file ID

**Comment by RyanCavanaugh** [comment]
Closing language service bugs related to the 6.0 implementation. For more information, see #62827
